016-database-adapters.md

```diff
@@ -0,0 +1,340 @@
+
+Write your answer below.
+
+-->
+
+### Utopia Database
+
+We'll implement an extra database abstraction layer as part of the Utopia project. This layer will include support for multiple database, and will expose a consistent API which will act as the lowest common denominator for all the supported adapters.
+
+The list of supported adapters will include:
+* Postgres
+* MySQL
+* MariaDB
+* MongoDB¹
+
+#### Data Types
+
+This library will support storing and fetching of all common JSON simple and complex [data types](https://restfulapi.net/json-data-types/).
+
+##### Simple Types
+
+* String
+* Integer
+* Float
+* Boolean
+* Null
+
+##### Complex Types
+* Array
+* Object
+* Relationships
+  * Reference (collection / document)
+  * References (Array of - collection / document)
+
+> Databases that don't support the storage of complex data types should store them as strings and parse them correctly when fetched.
+
+#### Persistency
+
+Each database adapter should support the following actions for fast storing and retrieval of collections of documents.
+
+**Databases** (Schemas for SQL)
+* create
+* delete
+
+**Collections** (Tables for SQL)
+* createCollection($name)
+* deleteCollection($name)
+
+**Attributes** (Table columns for SQL)
+* createAttribute(string $collection, string $name, string $type)
+* deleteAttribute(string $collection, string $name)
+
+**Indexes** (Table indexes for SQL)
+* createIndex(string $collection, string $name, string $type)
+* deleteIndex(string $collection, string $name, string $type)
+
+**Documents** (Table rows columns for SQL)
+* getDocument(string $collection, $id)
+* createDocument(string $collection, array $data)
+* updateDocument(string $collection, $id, array $data)
+* deleteDocument(string $collection, $id)
+
+#### Queries
+
+Each database adapter should allow for simple and advanced queries in consideration of underlying limitations.
+
+Method for querying data:
+* find(string $collection, $filters)
+* findFirst(string $collection, $filters) (extending `find`)
+* findLast(string $collection, $filters) (extending `find`)
+* count(string $collection, $filters) (extending `find`)
+
+##### Supported Operators
+* Equal (==)
+* Not Equal (!=)
+* Less Than (<)
+* Less or equal (<=)
+* Bigger Than (>)
+* Bigger or equal (>=)
+* Contains / In
+* Not Contains / Not In
+* Is Null
+* Is Not Null
+* Is Empty
+* Is Not Empty
+
+##### Joins / Relationships
+
+#### Paging
+
+Each database adapter should support two methods for paging. The first method is the classic `Limit and Offset`. The second method is `Limit and After` paging, which allows for better performance at a larger scale.
+
+#### Orders
+
+Support multi-column orders, multiple order type (ASC, DESC). We should leverage native casting for all types and avoid any manual casting of the data.
```

Comment:
Can we support Random order?
This is one of the biggest still existing issue with `Firebase` as well.